feat: aggregation mode explorer

MarcosNicolau opened the pull request:
## Description

Adds periodic fetches for aggregation mode events and stores data structures in db.  

## How to run
1. Build db: `make explorer_build_db`
2. Run db: `make explorer_run_db`
3. Run migration: `make explorer_ecto_setup_db`
4. Run explorer: `make run_explorer`
5. Verify the data is added via `psql`:
    - ` docker exec -it explorer-postgres-container psql -U explorer_user -d explorer_db`
    - `SELECT * FROM aggregated_proofs;`
    - `SELECT * FROM proofs_agg_mode WHERE aggregated_proof_number = N` 

> Note: to test this you'll need to setup two new ENV vars:
> - `ALIGNED_PROOF_AGG_CONFIG_FILE`: the config file for the aggregation service contract deployment 
> - `BEACON_CLIENT`: a beacon client for the network (public ones would not respond blob requests to me, so you'll have to get one for yourself probably).


## Type of change

- [x] New feature

## Checklist

- [ ] “Hotfix” to `testnet`, everything else to `staging`
- [ ] Linked to Github Issue
- [ ] This change depends on code or research by an external entity
  - [ ] Acknowledgements were updated to give credit
- [ ] Unit tests added
- [ ] This change requires new documentation.
  - [ ] Documentation has been added/updated.
- [ ] This change is an Optimization
  - [ ] Benchmarks added/run
- [ ] Has a known issue
  - [Link to the open issue addressing it]() 
- [ ] If your PR changes the Operator compatibility (Ex: Upgrade prover versions)
  - [ ] This PR adds compatibility for operator for both versions and do not change batcher/docs/examples
  - [ ] This PR updates batcher and docs/examples to the newer version. This requires the operator are already updated to be compatible

MarcosNicolau committed:
feat: aggregated-mode structure + silly sp1 program

MarcosNicolau committed:
feat: interfaces + basic service to run sp1

MarcosNicolau committed:
feat: sp1 aggregation program

MarcosNicolau committed:
refactor: rename libs and functions

MarcosNicolau committed:
feat: prove feature to not run prover locally

MarcosNicolau committed:
feat: proof aggregation service contract

MarcosNicolau committed:
feat: proof aggregation service deploy scripts

MarcosNicolau committed:
chore: update anvil state json

MarcosNicolau committed:
refactor: use mock client for mock proves

MarcosNicolau committed:
feat: verify proofs

MarcosNicolau committed:
refactor: folder structure

MarcosNicolau committed:
refactor: move files around more

MarcosNicolau committed:
feat: proof aggregator backend

MarcosNicolau committed:
feat: merkle tree hash with keccak

MarcosNicolau committed:
fix: write proofs to sp1 stdin

MarcosNicolau committed:
feat: verify supported proofs

MarcosNicolau committed:
feat: sign transaction + merkle root computation

MarcosNicolau committed:
feat: backend initialization

MarcosNicolau committed:
feat: load local proofs and add to proof aggregator

MarcosNicolau committed:
fix: skip when no proofs in queue

MarcosNicolau committed:
feat: set proof as missed on error

MarcosNicolau committed:
docs: aggregation mode readme

MarcosNicolau committed:
feat: read config from yaml file

MarcosNicolau committed:
chore: add config files + make commands

MarcosNicolau committed:
feat: s3 service

MarcosNicolau committed:
refactor: move queue to its own struct

MarcosNicolau committed:
feat: proofs fetcher

MarcosNicolau committed:
feat: main method initializer

MarcosNicolau committed:
chore: deps

MarcosNicolau committed:
feat: fetch logs instead of listening for events via ws

MarcosNicolau committed:
refactor: remove vk from sp1 proof

MarcosNicolau committed:
chore: add more tracing logs

MarcosNicolau committed:
feat: remove queue and returns proofs from fetcher

MarcosNicolau committed:
chore: update aligned service manager abi

MarcosNicolau committed:
feat: aggregated proof getter for ProofAggregationService

MarcosNicolau committed:
docs: update readme instructions

MarcosNicolau committed:
chore: update proof aggregator config

MarcosNicolau committed:
feat: get block number based on provided config param

MarcosNicolau committed:
feat: remove looping and run service only once

MarcosNicolau committed:
docs: update readme instructions

MarcosNicolau committed:
feat: send blob transaction

MarcosNicolau committed:
fix: start_proof_aggregator_local_with_proving command

MauroToscano committed:
Fix readme and makefile

MauroToscano committed:
Merge branch 'feat/aggregation-mode' of github.com:yetanotherco/aligned_layer into feat/aggregation-mode

MarcosNicolau committed:
feat: attach blob to contract transaction instead of being separate

MarcosNicolau committed:
feat: take blob versioned hash instead of blob tx in verify contract

MarcosNicolau committed:
feat(contracts): merge events into a single one with a new status field

MarcosNicolau committed:
feat: aligned proof aggregation model and contract interface

MarcosNicolau committed:
feat: [wip] periodically fetch from ProofAggregatorService events

MarcosNicolau committed:
Merge branch 'feat/aggregation-mode' into feat/aggregation-mode-explorer

MarcosNicolau committed:
feat: fetch logs from contract

MauroToscano committed:
Aggregation mode simplification (#1845)

MarcosNicolau committed:
Merge remote-tracking branch 'origin/feat/aggregation-mode' into feat/aggregation-mode-explorer

MarcosNicolau committed:
feat: beacon client to fetch blob data

MarcosNicolau committed:
fix: compilation warnings

MarcosNicolau committed:
chore: .env variables

MarcosNicolau committed:
feat: foreign key for agg mode proofs

MarcosNicolau committed:
feat: aggregated proofs migration

MarcosNicolau committed:
feat: eth client to get block by number

MarcosNicolau committed:
feat: blob decoding + parse events binaries

MarcosNicolau committed:
feat: periodic fetch decoded blob data

MarcosNicolau committed:
refactor: more idiomatic elixir and made panics explicit

MarcosNicolau committed:
fix: ecto schema association

MarcosNicolau committed:
chore: better tracing for agg periodic tasks

MarcosNicolau committed:
Merge remote-tracking branch 'origin/staging' into feat/aggregation-mode-explorer

MarcosNicolau committed:
fix merge

MarcosNicolau committed:
fix readme in merge

JuArce commented on explorer/lib/explorer/contract_managers/aligned_proof_aggregation_service.ex:
```diff
@@ -0,0 +1,128 @@
+      )
+
+    parent_beacon_block_hash = Map.get(block, "parentBeaconBlockRoot")
+    slot = Explorer.BeaconClient.get_block_slot!(parent_beacon_block_hash) + 1
```
This process could be improved using 

`curl http://<beacon_client>:5052/eth/v1/beacon/headers\?parent_root=0x1cfd5640adb8a1b82eba1f2a483b65a777de0371fb9c4b5286e57cfb7399dc01`

This way we do not need to add 1, which can be error prone and requires more knowledges of the blob retrieval process

MarcosNicolau replied:
Nicely spotted, solved in [54281dd](https://github.com/yetanotherco/aligned_layer/pull/1846/commits/54281ddf2d234b4fe46066ab4b318d213a601bc4).

MarcosNicolau committed:
refactor: get block header by hash

this prevents the situation where we would increase by one the slot from the parent but no block was proposed in the next slot

MarcosNicolau committed:
Merge branch 'staging' into feat/aggregation-mode-explorer

MarcosNicolau committed:
feat: update models to new contract version

MarcosNicolau committed:
feat: block_timestamp field for agg proof model

JuArce commented on explorer/lib/explorer/periodically.ex:
```diff
@@ -68,6 +73,64 @@ defmodule Explorer.Periodically do
   end
 
+  def handle_info(:aggregated_proofs, state) do
+    read_block_qty = 310
```
Add a comment with the meaning of this number. I think it is equivalent to a bit more than 1 hour

MarcosNicolau replied:
Solved in [981828d](https://github.com/yetanotherco/aligned_layer/pull/1846/commits/981828d354fef4e264284969715740a5a67be38d).

JuArce commented on explorer/lib/explorer/models/aggregation_mode_proof.ex:
```diff
@@ -0,0 +1,46 @@
+      nil ->
+        Explorer.Repo.insert(changeset)
+
+      existing_proof ->
```
i think it is more a general problem in the explorer, but in some moment we should improve logs

MarcosNicolau replied:
I agree, we should improve the logs + add a ci for formatting.

JuArce commented on explorer/lib/explorer/eth_client.ex:
```diff
@@ -0,0 +1,30 @@
+    eth_send("eth_getBlockByNumber", [block_number, false])
+  end
+
+  def eth_send(method, params, id \\ 1) do
```
This could be declared as private

MarcosNicolau replied:
Solved in [981828d](https://github.com/yetanotherco/aligned_layer/pull/1846/commits/981828d354fef4e264284969715740a5a67be38d).

JuArce commented on explorer/lib/explorer/beacon_client.ex:
```diff
@@ -0,0 +1,78 @@
+    beacon_get("/eth/v1/beacon/blob_sidecars/#{slot}")
+  end
+
+  def beacon_get(method) do
```
This could be declared as private

MarcosNicolau replied:
Solved in [981828d](https://github.com/yetanotherco/aligned_layer/pull/1846/commits/981828d354fef4e264284969715740a5a67be38d).

JuArce commented on explorer/priv/repo/migrations/20250407002009_create_aggregated_proofs.exs:
```diff
@@ -0,0 +1,32 @@
+
+  def change do
+    create table(:aggregated_proofs, primary_key: false) do
+      add(:merkle_root, :string, primary_key: true)
```
Since can exist multiple proofs aggregated with the same merkle root, it can not be used as primary key (it can not be unique)

JuArce replied:
ID should be uuid (:binary_id in ecto) https://hexdocs.pm/ecto/Ecto.Schema.html#content

MarcosNicolau replied:
Solved in [48be71c](https://github.com/yetanotherco/aligned_layer/pull/1846/commits/48be71c7eaf9fc99a4d7e660ee7e6f6d7bbc03b6).

JuArce reviewed (changes requested):
The most important change is related to the db scheme

MarcosNicolau committed:
feat: agg_proofs relation by uuid instead of merkle root

MarcosNicolau committed:
refactor: juli's comments

MarcosNicolau committed:
fix: edge case of repeated proofs for different proofs ids

JuArce committed:
chore: fetch aggregated proofs every 1 minute
